state-chain/pallets/cf-elections/src/electoral_systems/tests/delta_based_ingress.rs

```diff
@@ -0,0 +1,383 @@
+	pub static CHANNELS_CLOSED: RefCell<Vec<AccountId>> = const { RefCell::new(vec![]) };
+}
+
+type AccountId = u32;
+type Amount = u64;
+type BlockNumber = u64;
+
+struct MockIngressSink;
+impl IngressSink for MockIngressSink {
+	type Account = AccountId;
+	type Asset = Asset;
+	type Amount = Amount;
+	type BlockNumber = BlockNumber;
+	type DepositDetails = ();
+
+	fn on_ingress(
+		channel: Self::Account,
+		asset: Self::Asset,
+		amount: Self::Amount,
+		_block_number: Self::BlockNumber,
+		_details: Self::DepositDetails,
+	) {
+		AMOUNT_INGRESSED.with(|cell| {
+			let mut ingresses = cell.borrow_mut();
+			ingresses.push((channel, asset, amount));
+		});
+	}
+
+	fn on_ingress_reverted(channel: Self::Account, asset: Self::Asset, amount: Self::Amount) {
+		AMOUNT_REVERTED.with(|cell| {
+			let mut reverts = cell.borrow_mut();
+			reverts.push((channel, asset, amount))
+		});
+	}
+
+	fn on_channel_closed(channel: Self::Account) {
+		CHANNELS_CLOSED.with(|cell| {
+			let mut closed = cell.borrow_mut();
+			closed.push(channel);
+		});
+	}
+}
+
+type SimpleDeltaBasedIngress = DeltaBasedIngress<MockIngressSink, (), AccountId>;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Encode, Decode)]
+struct DepositChannel {
+	pub account: AccountId,
+	pub asset: Asset,
+	pub total_ingressed: Amount,
+	pub block_number: BlockNumber,
+	pub close_block: BlockNumber,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Encode, Decode)]
+struct DepositChannels(Vec<DepositChannel>);
+impl DepositChannels {
+	pub fn new(inner: Vec<DepositChannel>) -> Self {
+		Self(inner)
+	}
+
+	pub fn to_state(&self) -> BTreeMap<AccountId, ChannelTotalIngressedFor<MockIngressSink>> {
+		self.0
+			.clone()
+			.into_iter()
+			.map(|channel| {
+				(
+					channel.account,
+					ChannelTotalIngressed {
+						amount: channel.total_ingressed,
+						block_number: channel.block_number,
+					},
+				)
+			})
+			.collect::<BTreeMap<_, _>>()
+	}
+
+	pub fn to_state_map(
+		&self,
+	) -> BTreeMap<(AccountId, Asset), ChannelTotalIngressedFor<MockIngressSink>> {
+		self.0
+			.clone()
+			.into_iter()
+			.map(|channel| {
+				(
+					(channel.account, channel.asset),
+					ChannelTotalIngressed {
+						amount: channel.total_ingressed,
+						block_number: channel.block_number,
+					},
+				)
+			})
+			.collect::<BTreeMap<_, _>>()
+	}
+
+	pub fn to_properties(
+		&self,
+	) -> BTreeMap<
+		AccountId,
+		(OpenChannelDetailsFor<MockIngressSink>, ChannelTotalIngressedFor<MockIngressSink>),
+	> {
+		self.0
+			.clone()
+			.into_iter()
+			.map(|channel| {
+				(
+					channel.account,
+					(
+						OpenChannelDetails {
+							asset: channel.asset,
+							close_block: channel.close_block,
+						},
+						ChannelTotalIngressed {
+							amount: channel.total_ingressed,
+							block_number: channel.block_number,
+						},
+					),
+				)
+			})
+			.collect::<BTreeMap<_, _>>()
+	}
+}
+
+fn to_raw_state_map<K: Encode, V: Encode, M: IntoIterator<Item = (K, V)>>(
```

Comment:
I think it's better to decode the state in the utility and then we compare the typed data rather than the encoded, so think we should change the utility to pass through the decoded pre and post state.

Reply:
it turns out that returning the decoded version requires the Key to implement `Ord` trait, which is not required by the ElectoralSystem trait (and not implemented by the type used for DeltaBasedIngress for Solana: `(Account, Asset)`. 
I think this is why the code originally stores the encoded version of everything

Reply:
I managed to change this! it turned out I just needed to add the trait bounds in. 

I also made the StateMap<Vec, Vec> instead of <Vec, Option<Vec>> just to make everything easier